### ISSUE TITLE
[1.1] Get QP code path working

### ISSUE THREAD

**Issue by akshayka** [cvxgrp/cvxpy#816]
The QP code path does not work. Here is the output of running `nosetests cvxpy/tests/test_qp.py`

```
EEEE
======================================================================
ERROR: test_all_solvers (cvxpy.tests.test_qp.TestQp)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/Akshay/projects/cvxpy/cvxpy/tests/test_qp.py", line 73, in test_all_solvers
    self.quad_over_lin(solver)
  File "/Users/Akshay/projects/cvxpy/cvxpy/tests/test_qp.py", line 104, in quad_over_lin
    self.solve_QP(p, solver)
  File "/Users/Akshay/projects/cvxpy/cvxpy/tests/test_qp.py", line 69, in solve_QP
    return problem.solve(solver=solver_name, verbose=True)
  File "/Users/Akshay/projects/cvxpy/cvxpy/problems/problem.py", line 271, in solve
    return solve_func(self, *args, **kwargs)
  File "/Users/Akshay/projects/cvxpy/cvxpy/problems/problem.py", line 465, in _solve
    data, inverse_data = self._solving_chain.apply(self)
  File "/Users/Akshay/projects/cvxpy/cvxpy/reductions/chain.py", line 65, in apply
    problem, inv = r.apply(problem)
  File "/Users/Akshay/projects/cvxpy/cvxpy/reductions/matrix_stuffing.py", line 63, in apply
    new_obj, new_var, r = self.stuffed_objective(problem, extractor)
  File "/Users/Akshay/projects/cvxpy/cvxpy/reductions/qp2quad_form/qp_matrix_stuffing.py", line 49, in stuffed_objective
    P, q, r = extractor.quad_form(problem.objective.expr)
  File "/Users/Akshay/projects/cvxpy/cvxpy/utilities/coeff_extractor.py", line 151, in quad_form
    quad_forms)
  File "/Users/Akshay/projects/cvxpy/cvxpy/utilities/coeff_extractor.py", line 117, in extract_quadratic_coeffs
    P = c_part @ P
ValueError: matmul: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (n?,k),(k,m?)->(n?,m?) (size 2 is different from 1)

======================================================================
ERROR: Test solve parametric problem vs full problem
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/Akshay/projects/cvxpy/cvxpy/tests/test_qp.py", line 453, in test_parametric
    prob.solve(solver=solver)
  File "/Users/Akshay/projects/cvxpy/cvxpy/problems/problem.py", line 271, in solve
    return solve_func(self, *args, **kwargs)
  File "/Users/Akshay/projects/cvxpy/cvxpy/problems/problem.py", line 465, in _solve
    data, inverse_data = self._solving_chain.apply(self)
  File "/Users/Akshay/projects/cvxpy/cvxpy/reductions/chain.py", line 65, in apply
    problem, inv = r.apply(problem)
  File "/Users/Akshay/projects/cvxpy/cvxpy/reductions/matrix_stuffing.py", line 80, in apply
    Afull, bfull = extractor.affine(expr_list)
ValueError: too many values to unpack (expected 2)

======================================================================
ERROR: Test issue arising with square plus parameter.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/Akshay/projects/cvxpy/cvxpy/tests/test_qp.py", line 484, in test_square_param
    prob.solve()
  File "/Users/Akshay/projects/cvxpy/cvxpy/problems/problem.py", line 271, in solve
    return solve_func(self, *args, **kwargs)
  File "/Users/Akshay/projects/cvxpy/cvxpy/problems/problem.py", line 465, in _solve
    data, inverse_data = self._solving_chain.apply(self)
  File "/Users/Akshay/projects/cvxpy/cvxpy/reductions/chain.py", line 65, in apply
    problem, inv = r.apply(problem)
  File "/Users/Akshay/projects/cvxpy/cvxpy/reductions/matrix_stuffing.py", line 80, in apply
    Afull, bfull = extractor.affine(expr_list)
ValueError: too many values to unpack (expected 2)

======================================================================
ERROR: Test warm start.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/Akshay/projects/cvxpy/cvxpy/tests/test_qp.py", line 428, in test_warm_start
    result = prob.solve(warm_start=False)
  File "/Users/Akshay/projects/cvxpy/cvxpy/problems/problem.py", line 271, in solve
    return solve_func(self, *args, **kwargs)
  File "/Users/Akshay/projects/cvxpy/cvxpy/problems/problem.py", line 465, in _solve
    data, inverse_data = self._solving_chain.apply(self)
  File "/Users/Akshay/projects/cvxpy/cvxpy/reductions/chain.py", line 65, in apply
    problem, inv = r.apply(problem)
  File "/Users/Akshay/projects/cvxpy/cvxpy/reductions/matrix_stuffing.py", line 63, in apply
    new_obj, new_var, r = self.stuffed_objective(problem, extractor)
  File "/Users/Akshay/projects/cvxpy/cvxpy/reductions/qp2quad_form/qp_matrix_stuffing.py", line 49, in stuffed_objective
    P, q, r = extractor.quad_form(problem.objective.expr)
  File "/Users/Akshay/projects/cvxpy/cvxpy/utilities/coeff_extractor.py", line 151, in quad_form
    quad_forms)
  File "/Users/Akshay/projects/cvxpy/cvxpy/utilities/coeff_extractor.py", line 117, in extract_quadratic_coeffs
    P = c_part @ P
ValueError: matmul: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (n?,k),(k,m?)->(n?,m?) (size 200 is different from 1)

----------------------------------------------------------------------
Ran 4 tests in 0.027s

FAILED (errors=4)
```

**Comment by akshayka** [comment]
Many but not all of the QP tests fail at line 105 of `matrix_stuffing.py`:

```
# TODO QPs go here for constraints. Need to cast into right dimensions.
Afull, bfull = extractor.affine(expr_list)
```

@SteveDiamond , what needs to be done to address that TODO?

**Comment by akshayka** [comment]
42b916c addresses the TODO mentioned in the previous comment, and it also fixes dual variable retrieval.

Here's the output of test_qp.py now:
======================================================================
ERROR: Test solve parametric problem vs full problem
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/Akshay/projects/cvxpy/cvxpy/tests/test_qp.py", line 453, in test_parametric
    prob.solve(solver=solver)
  File "/Users/Akshay/projects/cvxpy/cvxpy/problems/problem.py", line 284, in solve
    return solve_func(self, *args, **kwargs)
  File "/Users/Akshay/projects/cvxpy/cvxpy/problems/problem.py", line 532, in _solve
    data, inverse_data = self._solving_chain.apply(self)
  File "/Users/Akshay/projects/cvxpy/cvxpy/reductions/chain.py", line 65, in apply
    problem, inv = r.apply(problem)
  File "/Users/Akshay/projects/cvxpy/cvxpy/reductions/solvers/conic_solvers/mosek_conif.py", line 174, in apply
    inv_data = {self.VAR_ID: problem.variables()[0].id,
TypeError: 'list' object is not callable
-------------------- >> begin captured stdout << ---------------------
[array(0.5), array(0.1)]
[array(0.5), array(0.09999993)]

--------------------- >> end captured stdout << ----------------------

======================================================================
FAIL: test_all_solvers (cvxpy.tests.test_qp.TestQp)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/Akshay/projects/cvxpy/cvxpy/tests/test_qp.py", line 77, in test_all_solvers
    self.quad_form(solver)
  File "/Users/Akshay/projects/cvxpy/cvxpy/tests/test_qp.py", line 154, in quad_form
    self.assertItemsAlmostEqual(z, var.value, places=4)
  File "/Users/Akshay/projects/cvxpy/cvxpy/tests/base_test.py", line 34, in assertItemsAlmostEqual
    self.assertAlmostEqual(a[i], b[i], places)
  File "/Users/Akshay/projects/cvxpy/cvxpy/tests/base_test.py", line 38, in assertAlmostEqual
    super(BaseTest, self).assertAlmostEqual(a, b, places=places)
AssertionError: -1.4543656745987648 != 635.1749306787997 within 4 places (636.6292963533984 difference)

----------------------------------------------------------------------
Ran 4 tests in 0.218s

FAILED (errors=1, failures=1)

**Comment by akshayka** [comment]
resolved by 147a578bd8455798f4c870226ab6bddd77226f7d
